File: lib/models/daily_log.dart
"""
DailyLog model mirroring backend schema
"""
class DailyLog {
  final String id;
  final String userId;
  final DateTime date;
  final int steps;
  final double? sleepHours;
  final int? sleepQuality;
  final int waterMl;
  final List<MealEntry>? meals;
  final String? mood;
  final String? notes;
  final int? caloriesConsumed;
  final int? caloriesBurned;

  const DailyLog({
    required this.id,
    required this.userId,
    required this.date,
    this.steps = 0,
    this.sleepHours,
    this.sleepQuality,
    this.waterMl = 0,
    this.meals,
    this.mood,
    this.notes,
    this.caloriesConsumed,
    this.caloriesBurned,
  });

  factory DailyLog.fromJson(Map<String, dynamic> j) => DailyLog(
    id: j['id'] as String,
    userId: j['user_id'] as String,
    date: DateTime.parse(j['date'] as String),
    steps: j['steps'] as int? ?? 0,
    sleepHours: (j['sleep_hours'] as num?)?.toDouble(),
    sleepQuality: j['sleep_quality'] as int?,
    waterMl: j['water_ml'] as int? ?? 0,
    meals: (j['meals'] as List?)?.map((m) => MealEntry.fromJson(m)).toList(),
    mood: j['mood'] as String?,
    notes: j['notes'] as String?,
    caloriesConsumed: j['calories_consumed'] as int?,
    caloriesBurned: j['calories_burned'] as int?,
  );

  Map<String, dynamic> toJson() => {
    'date': date.toIso8601String().split('T')[0],
    'steps': steps,
    'sleep_hours': sleepHours,
    'sleep_quality': sleepQuality,
    'water_ml': waterMl,
    'meals': meals?.map((m) => m.toJson()).toList(),
    'mood': mood,
    'notes': notes,
    'calories_consumed': caloriesConsumed,
    'calories_burned': caloriesBurned,
  };
}

class MealEntry {
  final String name;
  final String? time;
  final int? calories;
  final String? notes;

  const MealEntry({required this.name, this.time, this.calories, this.notes});

  factory MealEntry.fromJson(Map<String, dynamic> j) => MealEntry(
    name: j['name'] as String,
    time: j['time'] as String?,
    calories: j['calories'] as int?,
    notes: j['notes'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'name': name,
    'time': time,
    'calories': calories,
    'notes': notes,
  };
}

File: backend/schemas/daily_log.py
"""
Daily Log Pydantic schemas
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import date, datetime


class MealEntry(BaseModel):
    name: str
    time: Optional[str] = None
    calories: Optional[int] = None
    notes: Optional[str] = None


class DailyLogCreate(BaseModel):
    date: date
    steps: Optional[int] = Field(0, ge=0)
    sleep_hours: Optional[float] = Field(None, ge=0, le=24)
    sleep_quality: Optional[int] = Field(None, ge=1, le=10)
    water_ml: Optional[int] = Field(0, ge=0)
    meals: Optional[List[MealEntry]] = None
    mood: Optional[str] = Field(None, pattern="^(happy|neutral|sad|anxious|energetic|tired)$")
    notes: Optional[str] = None
    calories_consumed: Optional[int] = Field(None, ge=0)
    calories_burned: Optional[int] = Field(None, ge=0)


class DailyLogUpdate(BaseModel):
    steps: Optional[int] = Field(None, ge=0)
    sleep_hours: Optional[float] = Field(None, ge=0, le=24)
    sleep_quality: Optional[int] = Field(None, ge=1, le=10)
    water_ml: Optional[int] = Field(None, ge=0)
    meals: Optional[List[MealEntry]] = None
    mood: Optional[str] = Field(None, pattern="^(happy|neutral|sad|anxious|energetic|tired)$")
    notes: Optional[str] = None
    calories_consumed: Optional[int] = Field(None, ge=0)
    calories_burned: Optional[int] = Field(None, ge=0)


class DailyLogResponse(BaseModel):
    id: str
    user_id: str
    date: date
    steps: int
    sleep_hours: Optional[float]
    sleep_quality: Optional[int]
    water_ml: int
    meals: Optional[List[MealEntry]]
    mood: Optional[str]
    notes: Optional[str]
    calories_consumed: Optional[int]
    calories_burned: Optional[int]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DailyLogDateRange(BaseModel):
    start_date: date
    end_date: date